### PR TITLE
[WIP] Feat/transport agent prototype

### PR DESCRIPTION
**Pull request by litinsky**
1) Implemented RabbitMQ transport: agent communication with both channels and services;
2) Implemented async Agent with current pipeline;
3) Added service caller and service formatter entities;
4) Added channel connectors.

### CODE REVIEW

File: connectors/formatters/deeppavlov.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    if out:
+        service_responses = []
+
+        for response in data:
+            service_response = {}
+            service_response['tokens'] = response[0]
+            service_response['tags'] = response[1]
+            service_responses.append(service_response)
+
+        return service_responses
```
[the remark this answers is not in the document]
Comment (my-master) [reply connectors/formatters/deeppavlov.py]: From one hand, yes. From another one, seems better than defining different functions like `input_formatter`, `output_formatter` and the necessity to initialize both in the Agent config. Users on the first place, then the pureness of code.

File: core/config.py [hunk 2]
```diff
@@ -1,175 +1,45 @@
...
+        if not isinstance(services, list):
+            raise TypeError(ERR_MSG_TEMPLATE.format(type_error_text))
+
+        if not all(isinstance(service, str) for service in services):
+            raise TypeError(ERR_MSG_TEMPLATE.format(type_error_text))
+
+        if stage_name == 'skills' and not services:
+            raise ValueError('Skills pipeline stage should contain at least one service')
+
+        if stage_name in single_service_stages and len(services) > 1:
```
[the remark this answers is not in the document]
Comment (my-master) [reply core/config.py]: You are right. But we are going to change the static pipeline to a configurable one in different PR.

File: core/transport/connectors/rabbitmq.py [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+        logger.info(f'Queue: {in_queue_name} bound to routing key: {routing_key}')
+
+    async def _on_message_callback(self, message: IncomingMessage) -> None:
+        message_json = json.loads(message.body, encoding='utf-8')
+        message_to_channel: ToChannelMessage = ToChannelMessage.from_json(message_json)
+        await self._loop.create_task(self._to_channel_callback(message_to_channel.user_id, message_to_channel.response))
+        await message.ack()
+        logger.debug(f'Processed message to channel: {str(message_json)}')
+
+    async def send_to_agent(self, utterance: str, channel_id: str, user_id: str, reset_dialog: bool) -> None:
```
Comment (my-master) [inline core/transport/connectors/rabbitmq.py]: `channel_id` and `user_id` are the same

File: core/transport/messages.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        self.channel_id = channel_id
+        self.user_id = user_id
+        self.utterance = utterance
+        self.reset_dialog = reset_dialog
+
+
+def get_transport_message(message_json: dict) -> TMessageBase:
+    message_type = message_json['type']
+
+    if message_type == 'service_task':
```
[the remark this answers is not in the document]
Comment (litinsky) [reply core/transport/messages.py]: Refactored

File: core/transport/messages.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    def __init__(self, agent_name: str, channel_id: str, user_id: str, response: str) -> None:
+        self.type = self.__class__.type
+        self.agent_name = agent_name
+        self.channel_id = channel_id
+        self.user_id = user_id
+        self.response = response
+
+
+class FromChannelMessage(MessageBase):
+    type = 'from_channel_message'
```
[the remark this answers is not in the document]
Comment (litinsky) [reply core/transport/messages.py]: Fixed

File: core/transport/messages.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+
+TMessageBase = TypeVar('TMessageBase', bound='TaskBase')
+
+
+class MessageBase:
+    @classmethod
+    def from_json(cls, message_json) -> TMessageBase:
+        message_type = message_json.pop('type', None)
+        if message_type != cls.type:
```
[the remark this answers is not in the document]
Comment (litinsky) [reply core/transport/messages.py]: removed

File: core/transport/messages.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+class FromChannelMessage(MessageBase):
+    type = 'from_channel_message'
+    agent_name: str
+    channel_id: str
+    user_id: str
+    utterance: str
+    reset_dialog: bool
+
+    def __init__(self, agent_name: str, channel_id: str, user_id: str, utterance: str, reset_dialog: bool) -> None:
+        self.type = self.__class__.type
```
[the remark this answers is not in the document]
Comment (litinsky) [reply core/transport/messages.py]: For now we'll leave current stile, but in future we consider about refactoring validation wiht cupnproto.

File: core/transport/base.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    _channel_id: str
+    _on_channel_callback: Callable[[str, str, str, bool], Awaitable]
+
+    def __init__(self, config: dict, on_channel_callback: Callable[[str, str, str, bool], Awaitable]) -> None:
+        self._config = config
+        self._channel_id = self._config['channel']['id']
+        self._on_channel_callback = on_channel_callback
+
+    @abstractmethod
+    async def send_to_channel(self, user_id: str, response: str) -> None:
```
[the remark this answers is not in the document]
Comment (litinsky) [reply core/transport/base.py]: fixed

File: core/config.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (litinsky) [reply core/config.py]: This file will be removed in upcoming pull request"

File: core/config.py [hunk 9]
```diff
@@ -1,175 +1,45 @@
...
+
+    verify(config)
+    return config
+
+
+def verify(config: dict) -> None:
+    verify_pipeline(config)
+
+
+def verify_pipeline(config: dict) -> None:
```
[the remark this answers is not in the document]
Comment (litinsky) [reply core/config.py]: Ok, but this file will be removed in upcoming PR.

File: core/config.py [hunk 10]
```diff
@@ -1,175 +1,45 @@
...
+    pipeline = config['agent']['pipeline']
+    type_error_text = 'Pipeline elements should be of List[str] types'
+    single_service_stages = ['skill_selector', 'response_selector', 'response_formatter']
+
+    if not isinstance(pipeline, dict):
+        raise TypeError(ERR_MSG_TEMPLATE.format('Pipeline should be dict instance'))
+
+    if len(pipeline) == 0:
```
[the remark this answers is not in the document]
Comment (litinsky) [reply core/config.py]: Ok, but this file will be removed in upcoming PR.

File: connectors/formatters/deeppavlov.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (litinsky) [reply connectors/formatters/deeppavlov.py]: Leave current implementation until we make final decision.

**Comment by my-master** [comment]
Close as deprecated.

**Pull request closed without merging**